src/memory.rs

```diff
@@ -68,13 +78,25 @@ pub struct Memory<'a, 'tcx> {
 
 impl<'a, 'tcx> Memory<'a, 'tcx> {
     pub fn new(layout: &'a TargetDataLayout) -> Self {
-        Memory {
+        let mut mem = Memory {
             alloc_map: HashMap::new(),
             functions: HashMap::new(),
             function_alloc_cache: HashMap::new(),
-            next_id: AllocId(0),
+            next_id: AllocId(1),
             layout: layout,
-        }
+        };
+        // alloc id 0 is reserved for ZSTs, this is an optimization to prevent ZST
+        // (e.g. function pointers, (), [], ...) from requiring memory
+        let alloc = Allocation {
+            bytes: Vec::new(),
+            relocations: BTreeMap::new(),
+            undef_mask: UndefMask::new(0),
+        };
+        mem.alloc_map.insert(AllocId(0), alloc);
```

Comment:
We need this alloc. The `move_` function reads from zst pointers. We can get around this by adding a zst-check, but I don't think it's very relevant.